### PR TITLE
Make models in search space hub work with one-shot

### PR DESCRIPTION
**Pull request by ultmaster**
Most models in search space hub don't work well with one-shot strategy because lack of some features.

This PR makes most spaces work (at least runnable), except:

- NAS-Bench-101: the one-shot impl is a bit complex and I don't want to implement it yet.
- AutoFormer: still in stage 1.

This PR includes the following new features:

- Support mutable group in Conv2d for differentiable one-shot NAS.
- Implement a weighted-sum to sum complex types like nested list / dict.
- Extract a ``reduction()`` method for reduction customization (see following).
- Add a customized mutation for repeat in NDS spaces to support different output shape for different depths.
- Add a special rule for grouped Conv2d, when channels divided by groups is a constant, such that we don't allocate extra unused parameters.

It also contains some minor bug fixes.

A test suite is added. The tests take around 20 minutes on a 1080 GPU. It's not enabled on fast test yet.

### CODE REVIEW

File: nni/retiarii/oneshot/pytorch/supermodule/operation.py [hunk 1]
```diff
@@ -265,14 +266,18 @@ class MixedConv2d(MixedOperation, nn.Conv2d):
...
+    - ``groups``
     - ``stride`` (only supported in path sampling)
     - ``kernel_size``
-    - ``padding`` (only supported in path sampling)
+    - ``padding``
     - ``dilation`` (only supported in path sampling)
 
     ``padding`` will be the "max" padding in differentiable mode.
 
+    Mutable ``groups`` is NOT supported in most cases of differentaible mode.
```
Comment (QuanluZhang) [inline nni/retiarii/oneshot/pytorch/supermodule/operation.py:277]: -> differentiable

**Review by QuanluZhang — APPROVED** [review]

File: nni/retiarii/hub/pytorch/nasnet.py [hunk 2]
```diff
@@ -348,6 +351,100 @@ def __call__(self, repeat_idx: int):
...
+            return super().reduction(items, sampled)
+        # items[0] must be the result of first cell
+        assert len(items[0]) == 2
+        # Only apply the transformation on "prev" output.
+        items[0] = (self.first_cell_transformation(items[0][0]), items[0][1])
+        return super().reduction(items, sampled)
+
+
+class NDSStageDifferentiable(DifferentiableMixedRepeat):
+    """The differentiable implementation (for one-shot) of each NDS stage if depth is mutating."""
```
[the remark this answers is not in the document]
Comment (matluster) [reply nni/retiarii/hub/pytorch/nasnet.py:419]: Path-sampling: ENAS, SPOS, ...
Differentiable: DARTS, SNAS, FBNet, ProxylessNAS, ...

The difference is whether architecture weights are used.

File: nni/retiarii/hub/pytorch/proxylessnas.py [hunk 3]
```diff
@@ -133,7 +132,8 @@ def __init__(
             ConvBNReLU(in_channels, out_channels, kernel_size=1, norm_layer=norm_layer, activation_layer=nn.Identity)
         ]
         super().__init__(*simplify_sequential(blocks))
-        self.has_skip = stride == 1 and in_channels == out_channels
+        # NOTE: "is" is used here instead of "==" to avoid creating a new value choice.
```
[the remark this answers is not in the document]
Comment (matluster) [reply nni/retiarii/hub/pytorch/proxylessnas.py:135]: `ValueChoice([1,2,3]) == ValueChoice([2,3,4])` will create a new ValueChoice, which is true iff. two value choices choose the same value.

We could use techniques like "expression simplify" to simplify those expressions which must be true in all conditions. But I think that's a bit too complex.

[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/hub/pytorch/proxylessnas.py:135]: if two value choices `a` `b` have the same label, the `a is b` statement is True or False?

BTW, it would be better to add comments in a proper place to explain the behavior of `is` and `==` of value choice

[the remark this answers is not in the document]
Comment (matluster) [reply nni/retiarii/hub/pytorch/proxylessnas.py:135]: > if two value choices a b have the same label, the a is b statement is True or False?

No. `is` is reference equality. It only returns true when the underlying pointer points to the same memory address.

I think people will notice this when debugging. When they accidentally use `==` by mistake, it will raise an error here because it's found to be a value choice. `is`, on the other hand, is not what you will commonly do in daily life. It's just a hack here to support two types of channels: both int and value choice of int.

**Comment by liuzhe-lz** [comment]
I don't really get it.
Too much content to me.

**Review by liuzhe-lz — APPROVED** [review]
